providers/implementations/digests/blake2_prov.c

```diff
@@ -8,123 +8,181 @@
-}
-
-const OSSL_DISPATCH ossl_blake2b512_functions[] = {
-    {OSSL_FUNC_DIGEST_NEWCTX, (void (*)(void))blake2b512_newctx},
-    {OSSL_FUNC_DIGEST_UPDATE, (void (*)(void))ossl_blake2b_update},
-    {OSSL_FUNC_DIGEST_FINAL, (void (*)(void))blake2b512_internal_final},
-    {OSSL_FUNC_DIGEST_FREECTX, (void (*)(void))blake2b512_freectx},
-    {OSSL_FUNC_DIGEST_DUPCTX, (void (*)(void))blake2b512_dupctx},
-    {OSSL_FUNC_DIGEST_GET_PARAMS, (void (*)(void))blake2b512_get_params},
-    {OSSL_FUNC_DIGEST_GETTABLE_PARAMS,
-     (void (*)(void))ossl_digest_default_gettable_params},
-    {OSSL_FUNC_DIGEST_INIT, (void (*)(void))blake2b512_internal_init},
-    {OSSL_FUNC_DIGEST_GETTABLE_CTX_PARAMS,
-     (void (*)(void))ossl_blake2b_gettable_ctx_params},
-    {OSSL_FUNC_DIGEST_SETTABLE_CTX_PARAMS,
-     (void (*)(void))ossl_blake2b_settable_ctx_params},
-    {OSSL_FUNC_DIGEST_GET_CTX_PARAMS,
-     (void (*)(void))ossl_blake2b_get_ctx_params},
-    {OSSL_FUNC_DIGEST_SET_CTX_PARAMS,
-     (void (*)(void))ossl_blake2b_set_ctx_params},
-    {0, NULL}
+#define IMPLEMENT_BLAKE_functions(variant, VARIANT, variantsize) \
+static const OSSL_PARAM known_blake##variant##_ctx_params[] = { \
+    {OSSL_DIGEST_PARAM_SIZE, OSSL_PARAM_UNSIGNED_INTEGER, NULL, 0, 0}, \
+    OSSL_PARAM_END \
+}; \
+ \
+const OSSL_PARAM *ossl_blake##variant##_gettable_ctx_params(ossl_unused void *ctx, \
+                                                   ossl_unused void *pctx) \
+{ \
+    return known_blake##variant##_ctx_params; \
+} \
+ \
+const OSSL_PARAM *ossl_blake##variant##_settable_ctx_params(ossl_unused void *ctx, \
+                                                   ossl_unused void *pctx) \
+{ \
+    return known_blake##variant##_ctx_params; \
+} \
+ \
+int ossl_blake##variant##_get_ctx_params(void *vctx, OSSL_PARAM params[]) \
+{ \
+    struct blake##variant##_md_data_st *mdctx = vctx; \
+    OSSL_PARAM *p; \
+ \
+    BLAKE##VARIANT##_CTX *ctx = &mdctx->ctx; \
+ \
+    if (ctx == NULL) \
+        return 0; \
+    if (params == NULL) \
+        return 1; \
+ \
+    p = OSSL_PARAM_locate(params, OSSL_DIGEST_PARAM_SIZE); \
+    if (p != NULL \
+        && !OSSL_PARAM_set_uint(p, (unsigned int)mdctx->params.digest_length)) { \
+        ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER); \
+        return 0; \
+    } \
+ \
+    return 1; \
+} \
+ \
+int ossl_blake##variant##_set_ctx_params(void *vctx, const OSSL_PARAM params[]) \
+{ \
+    size_t size; \
+    struct blake##variant##_md_data_st *mdctx = vctx; \
+    const OSSL_PARAM *p; \
+ \
+    BLAKE##VARIANT##_CTX *ctx = &mdctx->ctx; \
+ \
+    if (ctx == NULL) \
+        return 0; \
+    if (params == NULL) \
+        return 1; \
+ \
+    p = OSSL_PARAM_locate_const(params, OSSL_DIGEST_PARAM_SIZE); \
+    if (p != NULL) { \
+        if (!OSSL_PARAM_get_size_t(p, &size)) { \
+            ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER); \
+            return 0; \
+        } \
+        if (size < 1 || size > UINT8_MAX) { \
```

Comment:
Is this upper constraint correct.. RFC 7693 seems to have 1 <= outlen <=64

Reply:
libb2 also refuses lengths longer than the hash, but the paper says "put 0xFFFFFFFF in there if you don't know" (![](https://github.com/openssl/openssl/assets/6709544/239befaf-2a57-4c9a-9630-7b0d9f32f235)) – I daren't have a take on this, I'm just adding backslashes.